.github/workflows/cachegrind.yml

```
@@ -0,0 +1,40 @@
+name: Valgrind Cachegrind
+
+on:
+  pull_request:
+    branches: [main]
+jobs:
+  valgrind:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Setup | Rust
+        uses: ATiltedTree/setup-rust@v1
+        with:
+          rust-version: stable
+
+      - name: Install Valgrind
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y valgrind
+
+      - name: Run Valgrind
+        run: |
+          cargo b -r -p revm-test --bin snailtracer
+          valgrind --tool=cachegrind target/release/snailtracer 2>&1 | tee cachegrind_results.txt
+
+      - name: Valgrind results
+        id: valgrind_results
+        run: |
+          contents=$(printf "%s" "$(head -c 64000 cachegrind_results.txt)")
+          # dump line by line into a file
+          printf "Valgrind Results:\n\n\`\`\`\n%s\n\`\`\`" "$contents" > results.md
+
+      - name: Comment on PR
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          gh pr comment ${{ github.event.pull_request.number }} --body-file results.md
```

```suggestion
          gh pr comment ${{ github.event.pull_request.number }} --edit-last --body-file results.md
```

Would add `--edit-last` and merge, we can debug live if needed.